apps/src/lib/node/ledger/protocol/transactions/ethereum_events/mod.rs

```diff
@@ -0,0 +1,393 @@
+        "got total voting power for epoch"
+    );
+
+    let voters = utils::get_voters_for_events(events.iter());
+    tracing::debug!(?voters, "Got validators who voted on at least one event");
+    let voting_powers =
+        utils::get_voting_powers_for_selected(&active_validators, voters)?;
+    tracing::debug!(
+        ?voting_powers,
+        "got voting powers for relevant validators"
+    );
+
+    let updates = events.into_iter().map(Into::<EthMsgUpdate>::into).collect();
+
+    let changed_keys =
+        apply_updates(store, updates, total_voting_power, voting_powers)?;
+
+    Ok(TxResult {
+        changed_keys,
+        ..Default::default()
+    })
+}
+
+/// Apply an Ethereum state update + act on any events which are confirmed
+pub(super) fn apply_updates(
+    store: &mut impl Store,
+    updates: HashSet<EthMsgUpdate>,
+    total_voting_power: VotingPower,
+    voting_powers: HashMap<Address, VotingPower>,
+) -> Result<BTreeSet<storage::Key>> {
+    tracing::debug!(
+        updates.len = updates.len(),
+        %total_voting_power,
+        ?voting_powers,
+        "Applying Ethereum state update transaction"
+    );
+
+    let mut changed_keys = BTreeSet::default();
+    let mut confirmed = vec![];
+    for update in updates {
+        // The order in which updates are applied to storage does not matter.
+        // The final storage state will be the same regardless.
+        let (mut changed, newly_confirmed) = apply_single_update(
+            store,
+            update.clone(),
+            total_voting_power,
+            &voting_powers,
+        )?;
+        changed_keys.append(&mut changed);
+        if newly_confirmed {
+            confirmed.push(update.body);
+        }
+    }
+    if confirmed.is_empty() {
+        tracing::debug!("No events were newly confirmed");
+        return Ok(changed_keys);
+    }
+    tracing::debug!(n = confirmed.len(), "Events were newly confirmed",);
+
+    // Right now, the order in which events are acted on does not matter.
+    // For `TransfersToNamada` events, they can happen in any order.
+    for event in &confirmed {
+        let mut changed = events::act_on(store, event)?;
+        changed_keys.append(&mut changed);
+    }
+    Ok(changed_keys)
+}
+
+/// Apply an [`EthMsgUpdate`] to storage. Returns any keys changed and whether
+/// the event was newly seen.
+fn apply_single_update(
+    store: &mut impl Store,
+    update: EthMsgUpdate,
+    total_voting_power: VotingPower,
+    voting_powers: &HashMap<Address, VotingPower>,
+) -> Result<(BTreeSet<storage::Key>, bool)> {
+    let eth_msg_keys = eth_msgs::Keys::from(&update.body);
+
+    // we arbitrarily look at whether the seen key is present to
+    // determine if the /eth_msg already exists in storage, but maybe there
+    // is a less arbitrary way to do this
+    let exists_in_storage = store.has_key(&eth_msg_keys.seen())?;
+
+    let (eth_msg, changed, newly_confirmed) = if !exists_in_storage {
```

Comment:
is it possible to split each of these branches of the if expr into two different helper funs?